### PR TITLE
Support range predicate pushdown for string columns with collation in PostgreSQL connector

### PR DESCRIPTION
**Pull request by takezoe**

### CODE REVIEW

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:516]: just `false`

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:511]: just `true`? 

ie what are PostgreSQL data types for which we would be here & `isCollatable`  would return false?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:873]: POSTGRESQL_CHARACTER_PUSHDOWN needs rename, since it's no longer what the name suggests it is

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1203]: drop `withCollate`, since it's (or: probably should be) always false

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1216]: is this line reachable?

File: testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java:39]:
```
SUPPORTS_AGGREGATION_PUSHDOWN_WITH_VARCHAR(fallback -> fallback.test(SUPPORTS_AGGREGATION_PUSHDOWN) && falback.test(SUPPORTS_PREDICATE_PUSHDOWN_WITH_VARCHAR_EQUALITY))
```

& move above `// Most connectors don't support` comment, since it's not applicable to this

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (findepi) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:287]: why still `isNotFullyPushedDown`?

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (findepi) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: retain TS.contraint assertion.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (findepi) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:845]: This is exact same as `varcharWriteFunctionWithCollate` (because we don't pad char values to their actual length)

you can call it `stringWriteFunctionWithCollate` and have common implementation

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1216]: When `jdbcTypeName` is `text`?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1203]: It can be true if `jdbcTypeName` is `text`.

File: testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java
Comment (takezoe) [reply testing/trino-testing/src/main/java/io/trino/testing/TestingConnectorBehavior.java]: Fixed.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: Fixed.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: How about `POSTGRESQL_STRING_PUSHDOWN_WITHOUT_COLLATE`?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: Ah, `CHAR` types can be always `true`. I dropped the `withCollate` parameter from `charColumnMapping()`.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:516]: `jdbcTypeName` can be `text`.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:287]: filterPredicate seems to remain on Trino side although the compacted predicate is pushed down to PostgreSQL:
```sql
SELECT "nationkey", "name", "regionkey" FROM "tpch"."nation" WHERE ("name" >= ? COLLATE "C" AND "name" <= ? COLLATE "C")
```
```
Output[regionkey, nationkey, name]
│   Layout: [regionkey:bigint, nationkey:bigint, name:varchar(25)]
└─ RemoteExchange[GATHER]
   │   Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint]
   └─ ScanFilter[table = postgresql:tpch.nation tpch.nation constraint on [name] columns=[nationkey:bigint:int8, name:varchar(25):varchar, regionkey:bigint:int8], filterPredicate = ("name" IN (CAST('POLAND' AS varchar(25)), CAST('ROMANIA' AS varchar(25)), CAST('VIETNAM' AS varchar(25))))]
          Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint]
          nationkey := nationkey:bigint:int8
          regionkey := regionkey:bigint:int8
          name := name:varchar(25):varchar
```

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:754]: Move this comment to the new method you added. `bpchar` isn't obvious and why these specific types are there isn't obvious otherwise.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:80]: Let's create a GitHub issue and add a TODO so that we can remove the additional behaviour once it's no longer needed (and other people know that this is something they can work on).

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: Just checking the existence of `FilterNode` is not enough as other many test cases that check whether predicate is not fully pushed down?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:754]: Moved.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:78]: Yeah, that makes sense (as long as this fix can be merged) and I would work on it once this pull request is completed. By the way, supporting type sensitive aggregation pushdown in JDBC plugins doesn't seem easy. Fundamental interface changes may be required.

Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:77]: Yeah, I ran into some similar issues in https://github.com/trinodb/trino/issues/7320.

There were some ideas floated in https://github.com/trinodb/trino/issues/7320#issuecomment-876981766 (which we didn't end up doing since it looked like a one-off need at that time).

If you already have some direction in your mind it might be helpful to discuss it on #dev on Slack too (if you think the changes will be large and touch the SPI).

Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:77]: Oh, will take a look at #7320 first. Thanks.

**Comment by takezoe** [comment]
@findepi Finished updating for now. Could you take a look again?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:898]: super duper nit: separate commit?

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (hashhar) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:76]: This is false only because you haven't changed the `implementJoin` code? Correct? We can possibly use the same technique to allow joins on varchars to get pushed down?

Asking because I'd like to submit a follow-up PR for the join pushdown case.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: I think this test included the TS.constraint because it was verifying that the connector didn't consume the filters which isn't the case anymore after this change (because the IN gets rewritten to a range predicate which now gets pushed down to the connector).

@takezoe If you revert this test change I assume the test fails?

**Review by hashhar — COMMENTED** [review]
Some minor comments.

Can we detect (or apply some heuristics) to determine during runtime whether a given predicate needs the COLLATION applied or not? For example for equality predicates adding the collation will lead to a definite performance regression because no indexes can be used anymore.

If it's not possible to do dynamically during runtime then let's add an `experimental.` prefixed config property to enable this behaviour and keep it opt in for now?

WDYT @wendigo ?

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: Oh nevermind, I misread the test. The "actual" predicate would still remain on the engine's side but the range predicate would get pushed down to connector.

Let's change the assertion to verify both assumptions - the tableScan has a range constraint applied and there is a filter node with the actual predicates applied.

Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: > Oh nevermind, I misread the test. The "actual" predicate would still remain on the engine's side but the range predicate would get pushed down to connector.

Yes, that's right.

> Let's change the assertion to verify both assumptions - the tableScan has a range constraint applied and there is a filter node above the tableScan.

Okay, let me try.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:898]: I will revert this change in this pull request.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:76]: Yes. This will be unnecessary if we implement join pushdown for VARCHAR columns.

**Comment by takezoe** [comment]
> For example for equality predicates adding the collation will lead to a definite performance regression because no indexes can be used anymore.

I think it's possible in some cases but 100% is impossible. But anyway it should be better than full-scan?

> If it's not possible to do dynamically during runtime then let's add an experimental. prefixed config property to enable this behaviour and keep it opt in for now?

I'm fine to make this optional as long as we can enable this by configuration.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: Fixed the test case.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: https://github.com/trinodb/trino/pull/9946

**Comment by hashhar** [comment]
> > For example for equality predicates adding the collation will lead to a definite performance regression because no indexes can be used anymore.
> 
> I think it's possible in some cases but 100% is impossible. But anyway it should be better than full-scan?
> 
> > If it's not possible to do dynamically during runtime then let's add an experimental. prefixed config property to enable this behaviour and keep it opt in for now?
> 
> I'm fine to make this optional as long as we can enable this by configuration. Is adding `experimental.postgresql.enable-string-pushdown-with-collate` ok?

@takezoe You can add a session property with matching config property in `PostgreSqlConfig`. I think `experimental.postgresql.string-pushdown-with-collate` is fine.

cc: @findepi @kokosing @wendigo @ebyhr Any opinions? I'd prefer to have this behind a configuration toggle (at-least for now) until we can verify that there isn't a performance concern in actual practical usage?

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (hashhar) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: Can you add a comment about why this assertion is testing in depth? I think we'll forget eventually.
Something like `Verify that a FilterNode is retained and only a compacted domain is pushed down to connector as a range predicate`.

**Review by hashhar — COMMENTED** [review]
Looks good % confirmation about config from other maintainers.

**Comment by takezoe** [comment]
Looks like `postgresql.experimental.string-pushdown-with-collate` is more consistent because `PostgreSqlConfig` used to have another experimental setting `postgresql.experimental.array-mapping`.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java
Comment (takezoe) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:294]: Added a comment.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1194]: Can we refactor to have different type-mappings installed/used depending on whether `isStringPushdownWithCollateEnabled` is true or false?

That way one of the impl will be the old "simple" mapping and a new impl will provide the mapping to be used when collation needs to be specified?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:238]: Maybe `isStringPushdownWithCollateEnabled`?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlConfig.java
Comment (hashhar) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlConfig.java:65]:
```suggestion
    @Config("postgresql.experimental.string-pushdown-with-collate.enabled")
```

since that seems to be convention for JdbcMetadataConfig and other recent configs introduced.

Comment (hashhar) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlConfig.java:65]: Adjust the getter and setter names accordingly.

**Review by hashhar — COMMENTED** [review]
LGTM % comments.

Mostly about possible simplifications.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java:1194]: Hmm, I see. Because I don't have enough time rest of this week and probably next week, I will suspend this pull request for a while and resume working when I can find another spaire time.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: Revised to `enableStringPushdownWithCollate`

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlConfig.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlConfig.java]: Revised to `postgresql.experimental.enable-string-pushdown-with-collate`

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java
Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: @hashhar What does `type-mappings` exactly mean here?

Comment (hashhar) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: I meant that the `toColumnMapping` and `toWriteMapping` methods can return different `ColumnMapping`/`WriteMapping`s depending on what the config is set to.

e.g. pseudo-java

```java
public WriteMapping toWriteMapping(ConnectorSession session, Type type)
{
    if (type instanceof VarcharType) {
        VarcharType varcharType = (VarcharType) type;
        String dataType;
        if (varcharType.isUnbounded()) {
            dataType = "varchar";
        }
        else {
            dataType = "varchar(" + varcharType.getBoundedLength() + ")";
        }
        
        if (isStringPushdownWithCollateEnabled) {
            return WriteMapping.sliceMapping(dataType, varcharWriteFunctionWithCollate());
        }

        return WriteMapping.sliceMapping(dataType, varcharWriteFunction());
     }
}
```

That way the actual `varcharWriteFunction` won't have too many conditionals and can be focused on doing just one thing (which it's name describes).

Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: Thanks for the explanation. Keeping many downstream objects simple and primitive by gathering complexity to a few upstream objects sounds a good trade!

Comment (takezoe) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java]: @hashhar Refactored. Do you think https://github.com/trinodb/trino/pull/9746/commits/1dc464865bc430fe7cc008b20b4e77a3903ba4e4 is better then the original code?

**Comment by takezoe** [comment]
@hashhar Finished updating the pull request. Could you take a look again?

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/PostgreSqlQueryRunner.java
Comment (hashhar) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/PostgreSqlQueryRunner.java:61]: Since you've now added a session property I suggest we add dedicated test method to TestPostgresqlConnectorTest which sets the session property and then verifies the result.

My reasoning is that tests should cover as much of "default" behaviour as possible too.

**Review by hashhar — COMMENTED** [review]
Looks good to me % comment. Thanks for working on this.

Also it looks the first two commits should be squashed together.

**Comment by takezoe** [comment]
Looks like `hasBehavior()` has to behave differently depending on whether `enable_string_pushdown_with_collate` is true or false, and it seems annoying to cover by a single test class. Does splitting `TestPostgreSqlConnectorTest` and create another test class just for testing string pushdown make sense?

**Comment by takezoe** [comment]
> having a flag as a field and changing its value inside `testStringPushdownWithCollate()`

I adopted this way for now and added a join on VARCHAR column case.

**Comment by hashhar** [comment]
> I adopted this way for now and added a case for join on VARCHAR columns.

Since only PostgreSQL has this behaviour today (and not enabled by default) I think it's fine to not add a new behaviour to `TestingConnectorBehaviour` just yet. That way you don't need the flag field (especially because we aren't using `hasBehaviour` in the new test at all).

In terms of future evolution I think once we prove it out with PostgreSQL we can lift the config to BaseJdbcConfig and then add a test in BaseJdbcConnectorTest which sets the session property - at that time we can revisit the best way to structure the `TestingConnectorBehaviour` (or maybe by that time we'll have more formal ways of handling collations).

@findepi @ebyhr @wendigo @takezoe Any thoughts?

**Comment by findepi** [comment]
@hashhar sounds good. except i am not convinced we need a config just yet, except for trying things out (a temporary kill switch)

**Comment by takezoe** [comment]
Ah, I see. That makes sense. I updated the test case.

**Comment by hashhar** [comment]
Closing and re-opening to get ci to run (some glitch on GitHub end).

**Review by hashhar — APPROVED** [review]

**Comment by hashhar** [comment]
Unrelated failure. Thanks @takezoe for the feature. Merging it.

**Pull request merged**
